Measure questions and either-neither

benadida opened the pull request.

benadida committed:
add support for labeled yes/no questions plus Ms either-or

benadida committed:
added to README for ms-either-or

benadida committed:
small field fix

benadida committed:
test coverage and relevant debugging

benadida committed:
no support for MsEitherNeither in v0

benadida committed:
ignore a couple of lines of coverage because we will never try to use v0 with mseitherneither

benadida committed:
fix README for rename of ms-either-or to ms-either-neither

eventualbuddha commented on src/election.ts:
```diff
@@ -197,6 +219,26 @@ export const getBallotStyle = ({
+ * Retrieve a contest from a set of contests based on ID
+ * special-cases Ms Either Neither contests
+ */
+const findContest = ({
```
```suggestion
export const findContest = ({
```

Maybe useful elsewhere?

eventualbuddha committed:
test(types): help TS with type inference

Using a type predicate helps TS refine the possible types in the `rawVote` properties.

eventualbuddha committed:
refactor: simplify contest finding code

We don't need to  check the type twice, we can just use a ternary.

eventualbuddha committed:
feat(types): expose `AnyContest` type

I often find myself writing `CandidateContest | YesNoContest`. With the addition of a third type, it's going to be painful without this.

eventualbuddha committed:
refactor: improve readability of roll-call code

- Use explicitly named properties for "either neither" and "pick one" contests.
- Use `readBoolean` instead of `readUint1` in boolean contexts.
- Correctly type the "either neither" and "pick one" votes as optional, since they might not be present.

eventualbuddha committed:
test: use the correct `ms-either-neither` name

eventualbuddha reviewed (approved):
I had a bunch of changes queued up as suggestions, but I figured it'd be better just to push them to the branch. Please do look them over. I had one other suggestion I didn't feel strongly about.

benadida committed:
export findContest

Co-authored-by: Brian Donovan <1938+eventualbuddha@users.noreply.github.com>

benadida commented:
> I had a bunch of changes queued up as suggestions, but I figured it'd be better just to push them to the branch. Please do look them over. I had one other suggestion I didn't feel strongly about.

All great changes. Thank you for making this code better :)
